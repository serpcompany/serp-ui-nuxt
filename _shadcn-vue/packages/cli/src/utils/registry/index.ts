import process from 'node:process'
import path from 'pathe'
import { HttpsProxyAgent } from 'https-proxy-agent'
import { ofetch } from 'ofetch'
import type * as z from 'zod'
import consola from 'consola'
import {
  registryBaseColorSchema,
  registryIndexSchema,
  registryWithContentSchema,
  stylesSchema,
} from '@/src/utils/registry/schema'
import type { registryItemWithContentSchema } from '@/src/utils/registry/schema'
import type { Config } from '@/src/utils/get-config'

const baseUrl = process.env.COMPONENTS_REGISTRY_URL ?? 'https://www.shadcn-vue.com'
const agent = process.env.https_proxy
  ? new HttpsProxyAgent(process.env.https_proxy)
  : undefined

export async function getRegistryIndex() {
  try {
    const [result] = await fetchRegistry(['index.json'])

    return registryIndexSchema.parse(result)
  }
  catch (error) {
    throw new Error('Failed to fetch components from registry.')
  }
}

export async function getRegistryStyles() {
  try {
    const [result] = await fetchRegistry(['styles/index.json'])

    return stylesSchema.parse(result)
  }
  catch (error) {
    throw new Error('Failed to fetch styles from registry.')
  }
}

export function getRegistryBaseColors() {
  return [
    {
      name: 'slate',
      label: 'Slate',
    },
    {
      name: 'gray',
      label: 'Gray',
    },
    {
      name: 'zinc',
      label: 'Zinc',
    },
    {
      name: 'neutral',
      label: 'Neutral',
    },
    {
      name: 'stone',
      label: 'Stone',
    },
  ]
}

export async function getRegistryBaseColor(baseColor: string) {
  try {
    const [result] = await fetchRegistry([`colors/${baseColor}.json`])

    return registryBaseColorSchema.parse(result)
  }
  catch (error) {
    throw new Error('Failed to fetch base color from registry.')
  }
}

export async function resolveTree(
  index: z.infer<typeof registryIndexSchema>,
  names: string[],
) {
  const tree: z.infer<typeof registryIndexSchema> = []

  for (const name of names) {
    const entry = index.find(entry => entry.name === name)

    if (!entry)
      continue

    tree.push(entry)

    if (entry.registryDependencies) {
      const dependencies = await resolveTree(index, entry.registryDependencies)
      tree.push(...dependencies)
    }
  }

  return tree.filter(
    (component, index, self) =>
      self.findIndex(c => c.name === component.name) === index,
  )
}

export async function fetchTree(
  style: string,
  tree: z.infer<typeof registryIndexSchema>,
) {
  try {
    const paths = tree.map(item => `styles/${style}/${item.name}.json`)
    const result = await fetchRegistry(paths)

    return registryWithContentSchema.parse(result)
  }
  catch (error) {
    throw new Error('Failed to fetch tree from registry.')
  }
}

export function getItemTargetPath(
  config: Config,
  item: Pick<z.infer<typeof registryItemWithContentSchema>, 'type'>,
  override?: string,
) {
  // Allow overrides for all items but ui.
  if (override)
    return override

  if (item.type === 'components:ui' && config.aliases.ui)
    return config.resolvedPaths.ui

  const [parent, type] = item.type.split(':')
  if (!(parent in config.resolvedPaths))
    return null

  return path.join(
    config.resolvedPaths[parent as keyof typeof config.resolvedPaths],
    type,
  )
}

async function fetchRegistry(paths: string[]) {
  try {
    const results = await Promise.all(
      paths.map(async (path) => {
        const response = await ofetch(`${baseUrl}/registry/${path}`, {
          // @ts-expect-error agent type
          agent,
        })

        return response
      }),
    )
    return results
  }
  catch (error) {
    consola.error(error)
    throw new Error(`Failed to fetch registry from ${baseUrl}.`)
  }
}
