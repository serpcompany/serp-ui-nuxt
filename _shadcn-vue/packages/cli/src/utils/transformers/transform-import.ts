import type { Transformer } from '@/src/utils/transformers'

export const transformImport: Transformer = async ({ sourceFile, config }) => {
  const importDeclarations = sourceFile.getImportDeclarations()

  for (const importDeclaration of importDeclarations) {
    const moduleSpecifier = importDeclaration.getModuleSpecifierValue()

    // Replace @/lib/registry/[style] with the components alias.
    if (moduleSpecifier.startsWith('@/lib/registry/')) {
      if (config.aliases.ui) {
        importDeclaration.setModuleSpecifier(
          moduleSpecifier.replace(/^@\/lib\/registry\/[^/]+\/ui/, config.aliases.ui),
        )
      }
      else {
        importDeclaration.setModuleSpecifier(
          moduleSpecifier.replace(
            /^@\/lib\/registry\/[^/]+/,
            config.aliases.components,
          ),
        )
      }
    }

    // Replace `import { cn } from "@/lib/utils"`
    if (moduleSpecifier === '@/lib/utils') {
      const namedImports = importDeclaration.getNamedImports()
      const cnImport = namedImports.find(i => i.getName() === 'cn')
      if (cnImport) {
        importDeclaration.setModuleSpecifier(
          moduleSpecifier.replace(/^@\/lib\/utils/, config.aliases.utils),
        )
      }
    }
  }

  return sourceFile
}
