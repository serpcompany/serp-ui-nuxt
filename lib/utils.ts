import { type ClassValue, clsx } from 'clsx';
import { twMerge } from 'tailwind-merge';

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function debounce<T extends (...args: unknown[]) => void>(
  callback: T,
  delay: number,
): (...args: Parameters<T>) => void {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  return (...args: Parameters<T>): void => {
    if (timeoutId) {
      clearTimeout(timeoutId);
    }

    timeoutId = setTimeout(() => {
      callback(...args);
    }, delay);
  };
}
