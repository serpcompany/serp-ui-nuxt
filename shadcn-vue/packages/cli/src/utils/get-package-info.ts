import { fileURLToPath } from 'node:url'
import path from 'pathe'
import fs from 'fs-extra'
import type { PackageJson } from 'type-fest'

export function getPackageInfo() {
  const packageJsonPath = getPackageFilePath('../package.json')

  return fs.readJSONSync(packageJsonPath) as PackageJson
}

function getPackageFilePath(filePath: string) {
  const distPath = fileURLToPath(new URL('.', import.meta.url))

  return path.resolve(distPath, filePath)
}
