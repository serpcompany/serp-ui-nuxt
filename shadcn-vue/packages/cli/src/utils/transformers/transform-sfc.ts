import { createRequire } from 'node:module'
import type { Transformer } from '@/src/utils/transformers'

// required cause Error: Dynamic require of "@babel/core" is not supported
const require = createRequire(import.meta.url)
const { transform } = require('detype')

export async function transformByDetype(content: string, filename: string) {
  return await transform(content, filename, {
    removeTsComments: true,
  })
}

export const transformSFC: Transformer<string> = async ({ sourceFile, config, filename }) => {
  const output = sourceFile?.getFullText()
  if (config?.typescript)
    return output

  return await transformByDetype(output, filename)
}
