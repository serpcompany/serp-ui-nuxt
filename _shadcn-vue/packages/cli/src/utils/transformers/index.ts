import { promises as fs } from 'node:fs'
import { tmpdir } from 'node:os'
import path from 'pathe'
import { Project, ScriptKind, type SourceFile } from 'ts-morph'
import type * as z from 'zod'
import type { Config } from '@/src/utils/get-config'
import type { registryBaseColorSchema } from '@/src/utils/registry/schema'
import { transformCssVars } from '@/src/utils/transformers/transform-css-vars'
import { transformImport } from '@/src/utils/transformers/transform-import'
import { transformSFC } from '@/src/utils/transformers/transform-sfc'

export interface TransformOpts {
  filename: string
  raw: string
  config: Config
  baseColor?: z.infer<typeof registryBaseColorSchema>
}

export type Transformer<Output = SourceFile> = (
  opts: TransformOpts & {
    sourceFile: SourceFile
  }
) => Promise<Output>

const transformers: Transformer[] = [
  transformCssVars,
  transformImport,
  // transformTwPrefixes,
]

const project = new Project({
  compilerOptions: {},
})

async function createTempSourceFile(filename: string) {
  const dir = await fs.mkdtemp(path.join(tmpdir(), 'shadcn-'))
  return path.join(dir, filename)
}

export async function transform(opts: TransformOpts) {
  const tempFile = await createTempSourceFile(opts.filename)
  const sourceFile = project.createSourceFile(tempFile, opts.raw, {
    scriptKind: ScriptKind.Unknown,
  })

  for (const transformer of transformers)
    transformer({ sourceFile, ...opts })

  return await transformSFC({
    sourceFile,
    ...opts,
  })
}
