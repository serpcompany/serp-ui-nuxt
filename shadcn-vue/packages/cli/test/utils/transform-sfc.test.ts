import { resolve } from 'pathe'
import { describe, expect, it } from 'vitest'
import { transform } from '../../src/utils/transformers'

describe('transformSFC', () => {
  it('basic', async () => {
    const result = await transform({
      filename: 'app.vue',
      raw: `<script lang="ts" setup>
      const array: (number | string)[] = [1, 2, 3]
      </script>
      
      <template>
        <div v-bind="{ array }">
          template
        </div>
      </template>
      
      <style scoped>
      </style>
      `,
      config: {},
    })
    expect(result).toMatchSnapshot()
  })

  it('remove all type reference', async () => {
    const result = await transform({
      filename: 'app.vue',
      raw: `<script lang="ts" setup>
      const array: (number | string)[] = [1, 2, 3]
      </script>
      
      <template>
        <div v-bind="{ array }" :prop="(a: number) => a" :prop2="(a: number) => {
          let b: number = a
          return b
        }">
          {{ true ? 123 as number : 0 }}
        </div>
      </template>
      
      <style scoped>
      </style>
      `,
      config: {},
    })
    expect(result).toMatchSnapshot()
  })

  it('defineProps', async () => {
    const result = await transform({
      filename: 'app.vue',
      raw: `<script lang="ts" setup>
      const props = defineProps<{ foo: string }>()
      </script>
      `,
      config: {},
    })
    expect(result).toMatchSnapshot()
  })

  it('defineProps with withDefaults', async () => {
    const result = await transform({
      filename: 'app.vue',
      raw: `<script lang="ts" setup>
      const props = withDefaults(defineProps<{ foo: string }>(), {
        foo: 'bar'
      })
      </script>
      `,
      config: {},
    })
    expect(result).toMatchSnapshot()
  })

  it('defineProps with external props', async () => {
    const result = await transform({
      filename: resolve(__dirname, './test.vue'),
      raw: `<script lang="ts" setup>
      import { type Props } from './__fixtures__/props'
      const props = withDefaults(defineProps<{ foo?: string } & Props>(), {
        foo: 'bar'
      })
      </script>
      `,
      config: {},
    })
    expect(result).toMatchSnapshot()
  })

  it('defineProps with package props', async () => {
    const result = await transform({
      filename: resolve(__dirname, './test.vue'),
      raw: `<script lang="ts" setup>
      import { type LabelProps } from 'radix-vue'
      const props = withDefaults(defineProps<{ foo?: string } & LabelProps>(), {
        foo: 'bar'
      })
      </script>
      `,
      config: {},
    })
    // TODO: We need to improve this. https://github.com/radix-vue/shadcn-vue/issues/187
    expect(result).toMatchSnapshot()
  })

  it('defineEmits', async () => {
    const result = await transform({
      filename: 'app.vue',
      raw: `<script lang="ts" setup>
      const emit = defineEmits<{ foo: string }>()
      </script>
      `,
      config: {},
    })
    expect(result).toMatchSnapshot()
  })
})
