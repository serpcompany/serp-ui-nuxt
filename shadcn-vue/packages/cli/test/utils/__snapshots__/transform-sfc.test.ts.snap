// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformSFC > basic 1`] = `
"<script setup>
const array = [1, 2, 3];
</script>

<template>
  <div v-bind="{ array }">template</div>
</template>

<style scoped></style>
"
`;

exports[`transformSFC > defineEmits 1`] = `
"<script setup>
const emit = defineEmits(["foo"]);
</script>
"
`;

exports[`transformSFC > defineProps 1`] = `
"<script setup>
const props = defineProps({
  foo: { type: String, required: true },
});
</script>
"
`;

exports[`transformSFC > defineProps with external props 1`] = `
"<script setup>
const props = defineProps({
  foo: { type: String, required: false, default: "bar" },
  a: { type: String, required: true },
  b: { type: Number, required: true },
});
</script>
"
`;

exports[`transformSFC > defineProps with package props 1`] = `
"<script setup>
const props = defineProps({
  foo: { type: String, required: false, default: "bar" },
  for: { type: String, required: false },
  asChild: { type: Boolean, required: false },
  as: { type: null, required: false },
});
</script>
"
`;

exports[`transformSFC > defineProps with withDefaults 1`] = `
"<script setup>
const props = defineProps({
  foo: { type: String, required: true, default: "bar" },
});
</script>
"
`;

exports[`transformSFC > remove all type reference 1`] = `
"<script setup>
const array = [1, 2, 3];
</script>

<template>
  <div
    v-bind="{ array }"
    :prop="(a) => a"
    :prop2="
      (a) => {
        let b = a;
        return b;
      }
    "
  >
    {{ true ? 123 : 0 }}
  </div>
</template>

<style scoped></style>
"
`;
